helix-core/src/comment.rs

```diff
@@ -94,6 +97,206 @@ pub fn toggle_line_comments(doc: &Rope, selection: &Selection, token: Option<&st
     Transaction::change(doc, changes.into_iter())
 }
 
+fn find_last_non_whitespace_char(text: RopeSlice) -> Option<usize> {
+    text.chars_at(text.len_chars())
+        .reversed()
+        .position(|ch| !ch.is_whitespace())
+        .map(|pos| text.len_chars() - pos - 1)
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct CommentChange {
+    range: Range,
+    start_pos: usize,
+    end_pos: usize,
+    start_margin: bool,
+    end_margin: bool,
+    start_token: String,
+    end_token: String,
+    commented: bool,
+}
+
+pub fn find_block_comments(
+    tokens: Vec<BlockCommentToken>,
+    text: RopeSlice,
+    selection: &Selection,
+) -> (bool, Vec<CommentChange>) {
+    let mut commented = true;
+    let mut comment_changes = Vec::with_capacity(selection.len());
+    let default_tokens = tokens.first().cloned().unwrap_or_default();
+    let mut start_token = default_tokens.start.clone();
+    let mut end_token = default_tokens.end.clone();
+
+    let mut tokens = tokens;
+    // sort the tokens by length, so longer tokens will match first
+    tokens.sort_by(|a, b| {
+        if a.start.len() == b.start.len() {
+            b.end.len().cmp(&a.end.len())
+        } else {
+            b.start.len().cmp(&a.start.len())
+        }
+    });
+    for range in selection {
+        let selection_slice = range.slice(text);
+        if let (Some(start_pos), Some(end_pos)) = (
+            find_first_non_whitespace_char(selection_slice),
+            find_last_non_whitespace_char(selection_slice),
+        ) {
+            let mut line_commented = false;
+            let mut after_start = 0;
+            let mut before_end = 0;
+
+            for BlockCommentToken { start, end } in &tokens {
+                let len = (end_pos + 1) - start_pos;
+                let start_len = start.chars().count();
```

Comment:
Tiny not, move this outside if the loop.

The conouker likely dies that anyway but I personally thjbj it makes the code easier to read if inky the stuff that actually needs to be calculated in a loop is inside the loop

Reply:
moved len outside the loop :+1: